[ews-build] Add steps to support filtering safer-cpp-results with results.webkit.org

briannafan opened the pull request:
#### ee9eeaa4c7f099c8ad6cc1f8ddbfe87346545d46
<pre>
[ews-build] Add steps to support filtering safer-cpp-results with results.webkit.org
<a href="https://bugs.webkit.org/show_bug.cgi?id=283064">https://bugs.webkit.org/show_bug.cgi?id=283064</a>
<a href="https://rdar.apple.com/139814797">rdar://139814797</a>

Reviewed by Jonathan Bedard.

These steps are needed after filtering safer-cpp-checks results using results database.
The buildbot master will filter results and modify unexpected_results.json file, so the
worker will need the updated file through DownloadUnexpectedResultsFromMaster.

With the updated results, we will need to re-generate the results index page using
GenerateSaferCPPResultsIndex.

Then, we want to delete full StaticAnalyzer results so they&apos;re not uploaded to
the master by using DeleteStaticAnalyzerResults.

(DownloadUnexpectedResultsFromMaster):
    - If results database filters out results, we need to move the new unexpected_results.json file
      over to the worker from the master.
    - This uses the pre-existing transfer.FileDownload step and is similar to DownloadBuiltProductFromMaster.

(DeleteStaticAnalyzerResults):
    - Added if there are unexpected results. We don&apos;t want to upload the entire 40GB+ results so
      this step deletes the specified StaticAnalyzer directory.

(GenerateSaferCPPResultsIndex):
    - If there are unexpected results after filtering, we need to generate a new results index page
      that displays the results with links to each report.
    - This is run on the worker and uses the unexpected_results.json file from DownloadUnexpectedResultsFromMaster.
    - It also sets the &apos;unexpected_new_issues&apos; property to be used by DisplaySaferCPPResults.

* Tools/CISupport/ews-build/steps_unittest.py: Added unittests for DownloadUnexpectedResultsFromMaster,
    DeleteStaticAnalyzerResults, and GenerateSaferCPPResultsIndex.

Canonical link: <a href="https://commits.webkit.org/287229@main">https://commits.webkit.org/287229@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a745c640b36d79b18d297fd24920fd420ea80103

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77848 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56881 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31219 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82496 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29105 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79969 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66039 "") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5175 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60959 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18903 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80916 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/66039 "") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66841 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41262 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/66039 "") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24310 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27448 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/66039 "") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24666 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83858 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5215 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3512 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69182 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/77525 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5371 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66761 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68431 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12444 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10545 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12173 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5163 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5155 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8587 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6940 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7352,6 +7352,95 @@ def getResultSummary(self):
+    def __init__(self, **kwargs):
+        # Allow the unit test to override mastersrc
+        if 'mastersrc' not in kwargs:
+            kwargs['mastersrc'] = self.mastersrc
```
Are any arguments other than `mastersrc` actually valid?

briannafan replied:
I took these from `DownloadBuiltProductFromMaster` - we could probably operate without them given the file is small, but I thought it would be good to be explicit, especially with the permissions (`mode=644`)

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7352,6 +7352,95 @@ def getResultSummary(self):
+
+    def run(self):
+        if not self.test:
+            self.mastersrc += f'/unexpected_results.json'
```
This is going to unconditionally add `/unexpected_results.json` every time `run()` is invoked. That's probably not what you intend. Under current logic, you won't ever run `run()` twice, but that's not an assumption you assert.

briannafan replied:
I just realized I can allow testing in a much cleaner way, updated in the next commit!

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7352,6 +7352,95 @@ def getResultSummary(self):
+        if rc != SUCCESS:
+            return defer.returnValue(rc)
+
+        logText = self.log_observer.getStdout()
```
How big are these logs? Can we easily fit the whole thing into memory? Should we doing the S3 uploading dance instead?

briannafan replied:
They're quite small! <100 lines, around 2KB, so I think it's fine :)

ryanhaddad commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7352,6 +7352,95 @@ def getResultSummary(self):
 
 
+class DownloadUnexpectedResultsFromMaster(transfer.FileDownload):
+    mastersrc = WithProperties('public_html/results/%(buildername)s')
```
Does this need to be updated given the changes to https://github.com/WebKit/WebKit/pull/36340/files#diff-32440cf8b81d39efcd9ad9036d550d72daef76fbff0f30dc3e0f85a415896990R7354?

briannafan committed:
[ews-build] Add steps to support filtering safer-cpp-results with results.webkit.org
https://bugs.webkit.org/show_bug.cgi?id=283064
rdar://139814797

Reviewed by Jonathan Bedard.

These steps are needed after filtering safer-cpp-checks results using results database.
The buildbot master will filter results and modify unexpected_results.json file, so the
worker will need the updated file through DownloadUnexpectedResultsFromMaster.

With the updated results, we will need to re-generate the results index page using
GenerateSaferCPPResultsIndex.

Then, we want to delete full StaticAnalyzer results so they're not uploaded to
the master by using DeleteStaticAnalyzerResults.

(DownloadUnexpectedResultsFromMaster):
    - If results database filters out results, we need to move the new unexpected_results.json file
      over to the worker from the master.
    - This uses the pre-existing transfer.FileDownload step and is similar to DownloadBuiltProductFromMaster.

(DeleteStaticAnalyzerResults):
    - Added if there are unexpected results. We don't want to upload the entire 40GB+ results so
      this step deletes the specified StaticAnalyzer directory.

(GenerateSaferCPPResultsIndex):
    - If there are unexpected results after filtering, we need to generate a new results index page
      that displays the results with links to each report.
    - This is run on the worker and uses the unexpected_results.json file from DownloadUnexpectedResultsFromMaster.
    - It also sets the 'unexpected_new_issues' property to be used by DisplaySaferCPPResults.

* Tools/CISupport/ews-build/steps_unittest.py: Added unittests for DownloadUnexpectedResultsFromMaster,
    DeleteStaticAnalyzerResults, and GenerateSaferCPPResultsIndex.

Canonical link: https://commits.webkit.org/287229@main
